polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -852,11 +920,48 @@ impl<Config: config::Config> XcmExecutor<Config> {
 					message_to_weigh.extend(xcm.0.clone().into_iter());
 					let (_, fee) =
 						validate_send::<Config::XcmSender>(dest.clone(), Xcm(message_to_weigh))?;
-					// set aside fee to be charged by XcmSender
-					let transport_fee = self.holding.saturating_take(fee.into());
-
+					let maybe_delivery_fee = if let Some(asset_needed_for_fees) = fee.get(0) {
+						log::trace!(
+							target: "xcm::DepositReserveAsset",
+							"Asset provided to pay for fees {:?}, asset required for transport fees: {:?}",
+							self.asset_used_for_fees, asset_needed_for_fees,
+						);
+						let asset_to_pay_for_fees =
+							self.asset_used_for_fees.as_ref().unwrap_or(&asset_needed_for_fees.id);
+						let actual_asset_to_use_for_fees =
```

Comment:
it would be nice if we can get rid of this repetition

Reply:
Once we get rid of jit withdrawal, we'll be able to just do the swap and be done with it.
Because we have jit withdrawal, we need to first quote to later withdraw the necessary assets for delivery fees.
By v5 we'll be able to remove this.